Updated source installation doc (#3389)
hyperspy/hyperspy: doc/user_guide/install.rst

.. Warning::

    When running hyperspy from a development version, it can happen that
    dependencies change when pulling new commits / syncing the project. You 
    will then need to manually update dependencies in the environment 
    (check `pyproject.toml <https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/pyproject.toml>`__) 
    or run again the development installation procedure, as explained below.

Installation in a Anaconda/Miniconda distribution
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Optionally, create an environment to separate your hyperspy installation from
other anaconda environments (`read more about environments here
<https://conda.io/projects/conda/en/latest/user-guide/concepts/environments.html>`_).
At this point, it is helpful to let conda pick compatible ``python`` and 
``hyperspy-dev`` versions. ``hyperspy-dev`` contains development 

> dependencies required for testing and building documentation. 

Reviewer: I would remove all mention of `hyperspy-dev`, because it is better to use the two environment file from the repository itself - these used by CI to setup the environment on Azure Pipelines.

```suggestion
```

.. code-block:: bash

>     $ conda create -n hspy_dev python hyperspy-dev -c conda-forge # create an env with compatible hyperspy-dev and python versions

Reviewer: ```suggestion
    $ conda env create -f conda_environment_dev.yml # create an env with development dependencies
```

    $ conda activate hspy_dev # activate environment

Install the runtime requirements using conda:

.. code-block:: bash

>     $ conda install hyperspy-base -c conda-forge --only-deps # install hyperspy dependencies

Reviewer: ```suggestion
    $ conda env update -f conda_environment.yml # install hyperspy dependencies
```

>     $ conda install hyperspy-base -c conda-forge --only-deps # install hyperspy dependencies

Reviewer 2: Not sure, if this line is still required then?

Author: I think so, otherwise you wouldn't get rosettasciio, scipy, etc. no? 

Since I don't know how to consult what's in hyperspy-dev I can't tell for sure (also not an expert), but I think the dependencies of hyperspy-base are not in there

    $ pip install -e . --no-deps # install the currently checked-out branch of hyperspy

> Note that some tests only pass on specific packages versions (see :ref:`Plot testing<plot-test-label>`).

Reviewer 2: ```suggestion
Note that some tests only pass on specific package versions (see :ref:`Plot testing<plot-test-label>`).
```

Consequently, a previous version of python (such as 3.10) can end up being 
installed in the ``hspy_dev`` environment. If this is a problem, ``hyperspy-dev`` should be 
removed from environment creation. This will ensure that the latest python release is used. 

Skipping ``hyperspy-dev`` will not prevent a successful installation from source, 
but will require manually installing some packages. Notably, ``gui-ipywidgets`` 
and ``gui-traitsui`` will be missing. More generally, all dependencies from ``dev`` 
will be absent (see :ref:`install-with-pip`). These can be installed manually 
using conda (inspect ``project.optional-dependencies`` section of 
`pyproject.toml <https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/pyproject.toml>`__ 

> and install them using ``conda install -c conda-forge``). Alternatively, all 
> optional dependencies installation can be left to ``pip``:

Reviewer 2: ```suggestion
and install them using ``conda install -c conda-forge``). Alternatively, the
installation of all optional dependencies can be left to ``pip``:
```

.. code-block:: bash

    $ conda create -n hspy_dev python # create an env with latest stable python
    $ conda activate hspy_dev # activate environment

>     $ conda install hyperspy-base -c conda-forge --only-deps # install hyperspy dependencies

Reviewer: Similarly as above, it would be better to use `conda env create -f conda_environment.yml`

    $ pip install -e .[dev] # Install hyperspy from source + dev dependencies (run from source folder)

Because this will result in many conflict risks being cleared by pip, this is not 

> necessarily the cleanest way to proceed.

Reviewer 2: ?

Author: The original install from source within a conda environment was failing (due to matplotlib not being forward-compatible with python 3.11+).  So I ended up modifying the guide to make it work again, but then you're stuck  with  python 3.10. 

This paragraph explain how to install from source on python 3.11+, both with and without pip for dependencies, and that there's a risk of failing the image tests
